quantization/src/encoded_vectors_pq.rs

```diff
@@ -0,0 +1,575 @@
+    /// * `encoded_vector` - Encoded result as a preallocated vector
+    fn encode_vector(
+        vector_data: &[f32],
+        vector_division: &[Range<usize>],
+        centroids: &[Vec<f32>],
+        encoded_vector: &mut Vec<u8>,
+    ) {
+        encoded_vector.clear();
+        for range in vector_division {
+            let subvector_data = &vector_data[range.clone()];
+            let mut min_distance = f32::MAX;
+            let mut min_centroid_index = 0;
+            for (centroid_index, centroid) in centroids.iter().enumerate() {
+                // because centroids are flattened by chunks, take centroid position using `range`
+                let centroid_data = &centroid[range.clone()];
+                // by product quantization algorithm use euclid metric for any similarity function
+                let distance = subvector_data
+                    .iter()
+                    .zip(centroid_data)
+                    .map(|(a, b)| (a - b).powi(2))
+                    .sum();
+                if distance < min_distance {
+                    min_distance = distance;
+                    min_centroid_index = centroid_index;
+                }
+            }
+            // encoding, replace whole chunk `range` by one `u8` index of nearest centroid
+            encoded_vector.push(min_centroid_index as u8);
+        }
+    }
+
+    /// Encode single vector from `&[f32]` into `&[u8]`.
+    /// This method divides `vector_data` into chunks, for each chunk
+    /// finds nearest centroid and replace whole chunk by nearest centroid index.
+    ///
+    /// # Arguments
+    /// * `data` - Original vector data
+    /// * `vector_division` - Division of original vector into chunks
+    /// * `vector_parameters` - parameters of original vector data (dimension, distance, ect)
+    /// * `centroids_count` - Count of centroids for each chunk
+    /// * `max_kmeans_threads` - Max allowed threads for kmeans process
+    fn find_centroids<'a>(
+        data: impl Iterator<Item = &'a [f32]> + Clone,
+        vector_division: &[Range<usize>],
+        vector_parameters: &VectorParameters,
+        centroids_count: usize,
+        max_kmeans_threads: usize,
+    ) -> Result<Vec<Vec<f32>>, EncodingError> {
+        let sample_size = KMEANS_SAMPLE_SIZE.min(vector_parameters.count);
+        let mut result = vec![vec![]; centroids_count];
+
+        // if there are not enough vectors, set centroids as point positions
+        if vector_parameters.count <= centroids_count {
+            for (i, vector_data) in data.into_iter().enumerate() {
+                result[i] = vector_data.to_vec();
+            }
+            for r in result
+                .iter_mut()
+                .take(centroids_count)
+                .skip(vector_parameters.count)
+            {
+                // fill empty centroids just with zeros
+                *r = vec![0.0; vector_parameters.dim];
+            }
```

Comment:
Range + [fill](https://doc.rust-lang.org/std/primitive.slice.html#method.fill):

```suggestion
            // fill empty centroids just with zeros
            result[vector_parameters.count..centroids_count].fill(vec![0.0; vector_parameters.dim]);
```